packages/common/test/index.test.js

```diff
@@ -866,95 +867,107 @@ describe('validate', () => {
+    const str = 'rock the cashbah';
+    const result = cursor(str)(state);
     expect(result.cursor).to.eql(str);
   });
 
   it('should clear the cursor', () => {
     const state = {
-      cursor: new Date()
-    }
-    const result = cursor()(state)
-    expect(result.cursor).to.eql(undefined)
+      cursor: new Date(),
+    };
+    const result = cursor()(state);
+    expect(result.cursor).to.eql(undefined);
   });
 
   it('should use a default value', () => {
-    const state = {}
-    const result = cursor(state.cursor, { defaultValue: 33 })(state)
-    expect(result.cursor).to.eql(33)
+    const state = {};
+    const result = cursor(state.cursor, { defaultValue: 33 })(state);
+    expect(result.cursor).to.eql(33);
   });
 
   it('should use a custom key', () => {
-    const state = {}
-    const result = cursor(44, { key: 'page' })(state)
-    expect(result.page).to.eql(44)
+    const state = {};
+    const result = cursor(44, { key: 'page' })(state);
+    expect(result.page).to.eql(44);
   });
 
   it('should re-use a custom key', () => {
-    const state = {}
-    const result1 = cursor(44, { key: 'page' })(state)
-    expect(result1.page).to.eql(44)
+    const state = {};
+    const result1 = cursor(44, { key: 'page' })(state);
+    expect(result1.page).to.eql(44);
 
-    const result2 = cursor(55)(state)
-    expect(result2.page).to.eql(55)
+    const result2 = cursor(55)(state);
+    expect(result2.page).to.eql(55);
   });
 
   // testing the log output is hard here, I've only verified it manally
   it('should use an object', () => {
-    const state = {}
-    const c = { page: 22, next: 23 }
-    const result = cursor(c, { key: 'cursor' })(state)
-    expect(result.cursor).to.eql(c)
+    const state = {};
+    const c = { page: 22, next: 23 };
+    const result = cursor(c, { key: 'cursor' })(state);
+    expect(result.cursor).to.eql(c);
   });
 
   it('should apply a custom formatter', () => {
     const state = {};
     const result = cursor('abc', {
-      format: (c) => c.toUpperCase()
+      format: c => c.toUpperCase(),
     })(state);
-    expect(result.cursor).to.eql('ABC')
-  })
+    expect(result.cursor).to.eql('ABC');
+  });
 
   it('should format "today"', () => {
     const state = {};
     const date = new Date().toDateString();
     const result = cursor('today', {
-      format: (c) => c.toDateString()
+      format: c => c.toDateString(),
     })(state);
-    expect(result.cursor).to.eql(date)
-  })
+    expect(result.cursor).to.eql(date);
+  });
 
   it('should format a number to an arbitrary object', () => {
     const state = {};
     const result = cursor(3, {
-      format: (c) => ({ page: c })
+      format: c => ({ page: c }),
     })(state);
-    expect(result.cursor).to.eql({ page: 3 })
-  })
+    expect(result.cursor).to.eql({ page: 3 });
+  });
+});
 
-});
\ No newline at end of file
+describe('group', () => {
+  it('should group an array of objects by a specified key path. ', async function () {
```

Comment:
As this is a common core function I'd like a few more tests please

- use a path rather than a key
- what happens if the key isn't present on any items (should happily return an empty object)
- what happens if a key value is undefined? eg, `{ city: undefined }`
- check that `arrayOfObjects` is expanded (pass a function for the first argument)
- check that `path` is expanded (pass a function for the second argument)

Also, no full stops at the end of test names please!

Reply:
I can but I'm not sure I understand the question! The path in your example looks fine?

Reply:
Ooh okay, i think we are align, i wanted to write a test for `use a path rather than a key` and i was confused on what this mean.

Reply:
> If the key isn't present on any items (should happily return an empty object)

This means were are removing any object that does not have that key right ?

Reply:
Well we don't _remove_ anything - we just don't add it. Basically `group` should always write an object to state.data, and that object may or may not have any groupings on it

Reply:
If we are not removing the stuff instead of empty object, shouldn't that be `undefined` key 
Eg:
```
const data = [
  { x: "a", y: { z: "a" } },
  { x: "b", y: { w: "a" } },
];

group(data, 'y.z');

//Should return 👇🏽 
{
  a: [ { x: 'a', y: { z 'a' }} ],
  undefined: [ { x: 'b', y: { w: 'a' } } ]
}
```

Reply:
I have update the tests, I am not sure if i am testing `expand reference` accurately would love your feedback